pkg/runner/healthcheck_common.go

```diff
@@ -0,0 +1,256 @@
+package runner
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"reflect"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/docker/go-units"
+	"github.com/ipfs/testground/pkg/api"
+	"github.com/ipfs/testground/pkg/docker"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type Checker func() bool
+type Fixer func() error
+
+// HealthcheckHelper is a strategy interface for runners.
+// Each runner may have required elements -- infrastructure, etc. which should be checked prior to
+// running plans. Individual checks are registered to the HealthcheckHelper using the Enlist()
+// method. With all of the checks enlisted, execute the checks, and optionally fixes, using the
+// RunChecks() method. The details of how the checks are performed is implementation specific.
+// Typically, the checker and fixer passed to the enlist method will be closures. These methods will
+// be called when RunChecks is executed.
+type HealthcheckHelper interface {
+	Enlist(name string, c Checker, f Fixer)
+	RunChecks(ctx context.Context, fix bool) error
+}
+
+type toDoElement struct {
+	Name    string
+	Checker Checker
+	Fixer   Fixer
+}
+
+// ErrGroupHealthcheckHelper implements HealthcheckHelper using an errgroup for paralellism.
+// WARNING: in order to prevent race conditions, the toDo slice should have an initial capacity of
+// one. See https://medium.com/@cep21/gos-append-is-not-always-thread-safe-a3034db7975
+type ErrgroupHealthcheckHelper struct {
+	toDo   []*toDoElement
+	report *api.HealthcheckReport
+}
+
+func (hh *ErrgroupHealthcheckHelper) Enlist(name string, c Checker, f Fixer) {
+	hh.toDo = append(hh.toDo, &toDoElement{name, c, f})
+}
+
+func (hh *ErrgroupHealthcheckHelper) RunChecks(ctx context.Context, fix bool) error {
+	eg, _ := errgroup.WithContext(ctx)
+	hh.report.Checks = make([]api.HealthcheckItem, len(hh.toDo))
+	hh.report.Fixes = make([]api.HealthcheckItem, len(hh.toDo))
+
+	for i, li := range hh.toDo {
+		i, li := i, li
+		hcp := *li
+		eg.Go(func() error {
```

Comment:
Parallel checks won't work. Some depend on others, e.g. the control network being available, so that we can start the sidecar, redis, etc. Let's run them sequentially, in the order they were enlisted. Parallelism won't buy us much anyway; the checks should be relatively lightweight.

Reply:
To express these dependencies, we might need to add a `dependsOn` parameter ([]string) when enlisting a check.

Reply:
That is a good point i hadn't thought of.

I was trying to speed up builds by enabling multiple-downloads, 

I hadn't encountered an error running it this way, but I can see that you are clearly right. The dependency graph is not addressed by starting all in parallel. Well, that's easy enough to change. I'll swap this out for a less complicated serial HealthcheckHelper.